lib/gpu/lib/src/render_pass.dart

```diff
@@ -0,0 +1,170 @@
+      this.resolveTexture = null});
+
+  LoadAction loadAction;
+  StoreAction storeAction;
+  ui.Color clearColor;
+  Texture texture;
+  Texture? resolveTexture;
+}
+
+base class StencilAttachment {
+  StencilAttachment(
+      {this.loadAction = LoadAction.clear,
+      this.storeAction = StoreAction.dontCare,
+      this.clearStencil = 0,
+      required this.texture});
+
+  LoadAction loadAction;
+  StoreAction storeAction;
+  int clearStencil;
+  Texture texture;
+}
+
+/// A descriptor for RenderPass creation. Defines the output targets for raster
+/// pipelines.
+base class RenderTarget {}
+
+base class RenderPass extends NativeFieldWrapperClass1 {
+  /// Creates a new RenderPass.
+  RenderPass._(CommandBuffer commandBuffer, ColorAttachment colorAttachment,
+      StencilAttachment? stencilAttachment) {
+    _initialize();
+    String? error;
+    error = _setColorAttachment(
+        colorAttachment.loadAction.index,
+        colorAttachment.storeAction.index,
+        colorAttachment.clearColor.value,
+        colorAttachment.texture,
+        colorAttachment.resolveTexture);
+    if (error != null) {
+      throw Exception(error);
+    }
+    if (stencilAttachment != null) {
+      error = _setStencilAttachment(
+          stencilAttachment.loadAction.index,
+          stencilAttachment.storeAction.index,
+          stencilAttachment.clearStencil,
+          stencilAttachment.texture);
+      if (error != null) {
+        throw Exception(error);
+      }
+    }
+    error = _begin(commandBuffer);
+    if (error != null) {
+      throw Exception(error);
+    }
+  }
+
+  void bindPipeline(RenderPipeline pipeline) {
+    _bindPipeline(pipeline);
+  }
+
+  void bindVertexBuffer(BufferView bufferView, int vertexCount) {
+    switch (bufferView.buffer.runtimeType) {
+      case DeviceBuffer:
+        _bindVertexBufferDevice(bufferView.buffer as DeviceBuffer,
+            bufferView.offsetInBytes, bufferView.lengthInBytes, vertexCount);
+        break;
+      case HostBuffer:
+        _bindVertexBufferHost(bufferView.buffer as HostBuffer,
+            bufferView.offsetInBytes, bufferView.lengthInBytes, vertexCount);
+        break;
+      default:
+        throw Exception("Invalid buffer type");
+    }
+  }
+
+  void bindUniform(UniformSlot slot, BufferView bufferView) {
+    bool success;
+    switch (bufferView.buffer.runtimeType) {
+      case DeviceBuffer:
+        success = _bindUniformDevice(
+            slot.shaderStage.index,
+            slot.slotId,
+            bufferView.buffer as DeviceBuffer,
+            bufferView.offsetInBytes,
+            bufferView.lengthInBytes);
+        break;
+      case HostBuffer:
+        success = _bindUniformHost(
+            slot.shaderStage.index,
+            slot.slotId,
+            bufferView.buffer as HostBuffer,
+            bufferView.offsetInBytes,
+            bufferView.lengthInBytes);
+        break;
+      default:
+        throw Exception("Invalid buffer type");
+    }
```

Comment:
Patterns like this indicate this should probably be a method on Buffer

Reply:
Normally I'd agree, but this needs to dispatch to different native methods tied to the `RenderPass` handle. So then `Buffer` would end up having something like a reverse `_bindTo(gpu.RenderPass, gpu.BufferView)` that then calls back into the right `RenderPass` native method, which feels a lot less straight forward.

Reply:
It's a visitor pattern 🙂

Reply:
Ok ok, I'll bite. Done.